### PR TITLE
Response stuffing

### PR DESCRIPTION
**Pull request by steveorsomethin**
This PR accomplishes two things:

It fixes a bug where whole branches were merged from jsonGraph responses, leading to nodes to be merged without also receiving their metadata such as $_absolutePath.

Adds a (for now) private model flag, _useServerPaths which can enable folks to opt into response stuffing. I opted for a private flag for now as I see this as a pretty major foot gun.

### CODE REVIEW

File: test/set/edge-cases.spec.js [hunk 1]
```diff
@@ -42,7 +42,7 @@ describe("Special Cases", function() {
         var edgeCaseCache = {
             jsonGraph: {
                 user: {
-                    name: "Jim",
+                    name: {$type: $atom, value: "Jim"},
```
Comment (steveorsomethin) [inline test/set/edge-cases.spec.js]: This test was actually testing for buggy behavior before. It was asserting that the name field wasn't getting wrapped, which was wrong.

File: lib/support/mergeJSONGraphNode.js [hunk 2]
```diff
@@ -192,7 +192,7 @@ module.exports = function mergeJSONGraphNode(
         }
     }
     else if (node == null) {
-        node = insertNode(message, parent, key, undefined, optimizedPath);
+        node = insertNode({}, parent, key, undefined, optimizedPath);
```
Comment (asyncanup) [inline lib/support/mergeJSONGraphNode.js]: i don't totally understand this part. what was `message` about, and why is it now replaced with an empty object?

**Comment by mattflix** [comment]
Is anybody looking at this PR?

The Windows apps team is waiting on it.

File: lib/request/GetRequestV2.js [hunk 3]
```diff
@@ -104,14 +104,21 @@ GetRequestV2.prototype = {
                             var currentVersion = incrementVersion.getCurrentVersion();
                             currentCacheVersion.setVersion(currentVersion);
                             var mergeContext = {hasInvalidatedResult : false};
-                            self._merge(rPaths, err, data, mergeContext);
+
+                            var pathsErr = model._useServerPaths && data.paths === undefined ?
+                                new Error("Server responses must include a 'paths' field when Model._useServerPaths === true") : undefined;
```
Comment (lrowe) [inline lib/request/GetRequestV2.js]: Is data guaranteed to be an object when err is an error?

File: lib/request/GetRequestV2.js [hunk 4]
```diff
@@ -190,7 +197,7 @@ GetRequestV2.prototype = {
         model._path = emptyArray;
 
         // flatten all the requested paths, adds them to the
-        var nextPaths = flattenRequestedPaths(requested);
+        var nextPaths = model._useServerPaths ? data.paths : flattenRequestedPaths(requested);
```
Comment (lrowe) [inline lib/request/GetRequestV2.js]: if requested paths need flattening might data.paths need flattening too?

File: lib/support/mergeJSONGraphNode.js (hunk 2 above)
Comment (lrowe) [reply lib/support/mergeJSONGraphNode.js]: It looks like message is the current JsonGraphNode of the response JsonGraph.

File: test/falcor/get/get.dataSource-only.spec.js [hunk 5]
```diff
@@ -275,6 +275,89 @@ describe('DataSource Only', function() {
...
+    it('should honor response-stuffed paths with _useServerPaths == true.', function(done) {
+        var onGet = sinon.spy();
+        var source = new LocalDataSource(cacheGenerator(0, 2), {
+            onGet: onGet,
+            wait: 100,
+            onResults: function(data) {
+                data.paths = [
+                    ['videos', 0, 'title'],
+                    ['videos', 1, 'title']
+                ];
```
Comment (lrowe) [inline test/falcor/get/get.dataSource-only.spec.js]: I imagine this might be collapsed server side to `[['videos', { from: 0, to: 1 }, 'title']]`.

File: lib/support/mergeJSONGraphNode.js (hunk 2 above)
Comment (jhusain) [reply lib/support/mergeJSONGraphNode.js]: This change creates a new node instead of assuming that the node from the server is correctly formed. This prevents extra data stuffed into a node from the server from piggybacking into the cache.

File: lib/request/GetRequestV2.js (hunk 4 above)
Comment (steveorsomethin) [reply lib/request/GetRequestV2.js]: It doesn't, only because it's really just removing an outer wrapped on requested that shouldn't be present on the envelope paths.

File: test/falcor/get/get.dataSource-only.spec.js (hunk 5 above)
Comment (steveorsomethin) [reply test/falcor/get/get.dataSource-only.spec.js]: Not a problem, as tons of tests already cover the case of complex keys (and they remain complex keys in _merge).

File: lib/request/GetRequestV2.js (hunk 3 above)
Comment (steveorsomethin) [reply lib/request/GetRequestV2.js]: Will add a null check just in case.

File: lib/support/mergeJSONGraphNode.js (hunk 2 above)
Comment (asyncanup) [reply lib/support/mergeJSONGraphNode.js]: perfect, this solves the issue: "model.set() claims object supplied in arguments" as well!
